core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -605,36 +691,42 @@ class KafkaController(val config: KafkaConfig,
         val topic = tp.topic
         val assignedReplicas = controllerContext.partitionReplicaAssignment(tp)
         if (assignedReplicas.nonEmpty) {
+          // TODO: Do we need  `&& reassignedPartitionContext.ongoingReassignmentOpt.isEmpty` to the if check here?
           if (assignedReplicas == newReplicas) {
             info(s"Partition $tp to be reassigned is already assigned to replicas " +
               s"${newReplicas.mkString(",")}. Ignoring request for partition reassignment.")
             partitionsToBeRemovedFromReassignment.add(tp)
+            reassignmentResults.put(tp, ApiError.NONE)
           } else {
             try {
+              // STAN: maybe print out previous assignment?
               info(s"Handling reassignment of partition $tp to new replicas ${newReplicas.mkString(",")}")
               // first register ISR change listener
               reassignedPartitionContext.registerReassignIsrChangeHandler(zkClient)
               // mark topic ineligible for deletion for the partitions being reassigned
               topicDeletionManager.markTopicIneligibleForDeletion(Set(topic),
                 reason = "topic reassignment in progress")
               onPartitionReassignment(tp, reassignedPartitionContext)
+              reassignmentResults.put(tp, ApiError.NONE)
             } catch {
               case e: ControllerMovedException =>
                 error(s"Error completing reassignment of partition $tp because controller has moved to another broker", e)
                 throw e
               case e: Throwable =>
                 error(s"Error completing reassignment of partition $tp", e)
-                // remove the partition from the admin path to unblock the admin client
                 partitionsToBeRemovedFromReassignment.add(tp)
+                reassignmentResults.put(tp, new ApiError(Errors.UNKNOWN_SERVER_ERROR))
             }
           }
         } else {
             error(s"Ignoring request to reassign partition $tp that doesn't exist.")
             partitionsToBeRemovedFromReassignment.add(tp)
+            reassignmentResults.put(tp, new ApiError(Errors.UNKNOWN_TOPIC_OR_PARTITION, "The partition does not exist."))
```

Comment:
Do we want to guarantee that if controllerContext.partitionReplicaAssignment(tp).isBeingReassigned then controllerContext.partitionsBeingReassigned(tp).isPresent?
At this point, I don't know if we should update the in-memory and ZK replica assignment to one that is not being reassigned. Since we remove the partition from the partitionsBeingReassigned collection, it seems to make sense to also update in-memory and ZK to truly treat the reassignment as failed.
Yet, do we need to send LeaderAndIsr requests if we change those?

The one scenario where this divergence could happen (controllerContext.partitionReplicaAssignment(tp).isBeingReassigned is true and controllerContext.partitionsBeingReassigned(tp).isPresent is not)
is if we have an UNKNOWN_SERVER_ERROR during the reassignment (line 723)

Reply:
I think we should try to avoid having an in-memory state that does not match the ZK state.  Can you think of a case where that might happen here?

Reply:
I think that might happen if we throw an error somewhere inside `onPartitionReassignment` after it updates the in-memory state. This wasn't handled previously either

The question here is more about the two separate in-memory collections diverging - `partitionsBeingReassigned` and `partitionReplicaAssignment`. If we get an unknown error inside `onPartitionReassignment` we may have updated the in-memory `partitionReplicaAssignment` but in `maybeTriggerPartitionReassignment` we catch the error and remove the partition from `partitionsBeingReassigned`. So we may end up with an in-memory state indicating a reassignment but none scheduled due to some failure.

I can't think of a great way to guard against this. It may be good to leave it as is

Reply:
It's difficult to handle arbitrary exceptions propagating from any line of code.  If we're confident that we've carefully checked the code in onPartitionReassignment, let's not block on this for now.